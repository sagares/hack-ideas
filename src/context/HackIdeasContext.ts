import { createContext } from "react";
import { IdeasArray } from "../HackIdea.types";

type Employee = {
  empId: string;
};

type HackIdeasContextType = {
  currentUser: Employee;
  logout: () => void;
  setCurrentUser: (employee: Employee) => void;
  ideas: IdeasArray;
  setIdeas: (ideas: IdeasArray) => void;
};

const HackIdeasContext = createContext<Partial<HackIdeasContextType>>({});

export { HackIdeasContext };
