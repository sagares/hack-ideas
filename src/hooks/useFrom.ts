import { useState } from "react";

function useForm<T>(initialState = {}, validations = []) {
  const [values, setValues] = useState<T>(initialState as T);
  const [errors, setErrors] = useState<T>({} as T);
  const [isValid, setIsValid] = useState(false);
  const [touched, setTouched] = useState<T>({} as T);

  const changeHandler = (event) => {
    let value: string | Array<string>;
    if (event.target.tagName === "SELECT") {
      value = Array.from(
        event.target.selectedOptions,
        (option: HTMLOptionElement) => option.value
      );
    } else {
      value = event.target.value;
    }
    const newValues = {
      ...values,
      [event.target.name]: value,
    };
    const { isValid, errors } = validate(validations, newValues);
    setValues(newValues);
    setIsValid(isValid);
    setErrors(errors);
    setTouched({ ...touched, [event.target.name]: true });
  };

  return { values, changeHandler, isValid, errors, touched };
}

function validate(validations, values) {
  const errors = validations
    .map((validation) => validation(values))
    .filter((validation) => typeof validation === "object");
  return {
    isValid: errors.length === 0,
    errors: errors.reduce((errors, error) => ({ ...errors, ...error }), {}),
  };
}

export default useForm;
