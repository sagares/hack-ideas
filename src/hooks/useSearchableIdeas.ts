import { useMemo } from "react";
import { IdeasArray } from "../HackIdea.types";

const useSearchableIdeas = (
  ideas: IdeasArray,
  searchText: string,
  sortBy: string,
  orderAsc: boolean
) => {
  return useMemo(() => {
    const escapedSearchString = searchText.replace(
      /[-[\]{}()*+?.,\\^$|#\s]/g,
      "\\$&"
    );
    const matchString = new RegExp(escapedSearchString, "gi");

    const filteredIdeas =
      ideas &&
      Object.keys(ideas).reduce((acc, key) => {
        const idea = ideas[key];
        if (
          idea.name.toLowerCase().match(matchString) ||
          idea.id.toLowerCase().match(matchString) ||
          idea.tags.includes(searchText)
        ) {
          acc.push(idea);
        }
        return acc;
      }, []);

    if (filteredIdeas) {
      if (sortBy === "submittedOn") {
        filteredIdeas.sort((idea1, idea2) => {
          return orderAsc ? idea2.submittedOn - idea1.submittedOn : idea1.submittedOn - idea2.submittedOn;
        });
      } else if (sortBy === "upvotes") {
        filteredIdeas.sort((idea1, idea2) => {
          const upvotes1 = idea1.upvotes?.length || 0;
          const upvotes2 = idea2.upvotes?.length || 0;
          return orderAsc ? upvotes2 - upvotes1 : upvotes1 - upvotes2;
        });
      }
    }

    return filteredIdeas;
  }, [ideas, searchText, sortBy, orderAsc]);
};

export default useSearchableIdeas;
