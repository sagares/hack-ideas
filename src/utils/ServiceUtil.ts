// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getDatabase, ref, onValue, set } from "firebase/database";
import {  Idea } from "../HackIdea.types";

const firebaseConfig = {
  apiKey: "AIzaSyC0RFn0RNnX8EoPHpAu-lf4EvN08aPKMCg",
  authDomain: "hack-ideas-c7cd7.firebaseapp.com",
  databaseURL: "https://hack-ideas-c7cd7-default-rtdb.firebaseio.com",
  projectId: "hack-ideas-c7cd7",
  storageBucket: "hack-ideas-c7cd7.appspot.com",
  messagingSenderId: "197788164361",
  appId: "1:197788164361:web:36a94bf614b4b5c830f3d8",
  measurementId: "G-660ZLLJJEB",
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const database = getDatabase();

const registerEmployee = (empId: string) => {
  set(ref(database, "employees/" + empId), {
    empId,
  });
};

const getEmployeeById = (empId: string, callback: Function) => {
  let empRef = ref(database, "employees/" + empId);
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getEmployeeData = (callback: Function) => {
  let empRef = ref(database, "employees/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getIdeas = (callback: Function) => {
  let empRef = ref(database, "ideas/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const registerIdea = (idea: Idea) => {
  const id = toCamelCase(idea.name);
  set(ref(database, "ideas/" + id), {
    ...idea,
    id,
    upvotes: [],
  });
};

const updateIdea = (ideaId: string, updatedIdea: Idea) => {
  set(ref(database, "ideas/" + ideaId), updatedIdea);
};

const toCamelCase = (sentenceCase: string) => {
  var out = "";
  sentenceCase.split(" ").forEach((el, idx) => {
    var add = el.toLowerCase();
    out += idx === 0 ? add : add[0].toUpperCase() + add.slice(1);
  });
  return out;
};

export {
  getEmployeeById,
  getEmployeeData,
  getIdeas,
  registerEmployee,
  registerIdea,
  updateIdea,
};
