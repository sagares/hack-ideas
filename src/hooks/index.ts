import useForm from "./useFrom";
import useFormInput from "./useFormInput";
import useSearchableIdeas from "./useSearchableIdeas";
import useSessionStorage from "./useSessionStorage";

export {
  useForm,
  useFormInput,
  useSearchableIdeas,
  useSessionStorage,
};
