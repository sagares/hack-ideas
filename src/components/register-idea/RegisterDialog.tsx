import React, { useContext, useEffect } from "react";
import M from "materialize-css";
import { RegisterDialogProps } from "../../HackIdea.types";
import { registerIdea } from "../../utils/ServiceUtil";
import HackIdeasContext from "../../context";
import { useForm } from "../../hooks";

type RegisterFormFields = {
  title: string;
  description: string;
  tags: Array<string>;
};

const RegisterDialog = ({ isOpen, onCancel }: RegisterDialogProps) => {
  const { currentUser } = useContext(HackIdeasContext);

  const isRequired = (value: string) => {
    return value != null && value.trim().length > 0;
  };

  const validations = [
    ({ title }) => isRequired(title) || { title: "Title is required" },
    ({ description }) =>
      isRequired(description) || { description: "Description is required" },
    ({ tags }) => tags.length > 0 || { tags: "One more tags required." },
  ];

  const { values, changeHandler, isValid, errors, touched } =
    useForm<RegisterFormFields>(
      {
        title: "",
        description: "",
        tags: [],
      },
      validations
    );

  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    registerIdea({
      name: values.title,
      description: values.description,
      tags: values.tags,
      submittedBy: currentUser.empId,
      submittedOn: Date.now(),
    });
    M.toast({ html: "Idea registered!!" });
    onCancel(false);
  };

  useEffect(() => {
    const selects = document.querySelectorAll("select");
    M.FormSelect.init(selects, {});
  }, []);

  return (
    <div className={`register-idea--modal ${isOpen ? "open" : ""}`}>
      <div id="modal1" className="modal modal-fixed-footer">
        <div className="modal-content">
          <h4>Register New Idea</h4>
          <p>Please fill in the form below to register your hack idea.</p>
          <div className="row">
            <form className="col s12">
              <div
                className={`title-field input-field col s12 ${
                  touched.title && errors.title ? "invalid-field" : ""
                }`}
              >
                <input
                  id="idea_title"
                  type="text"
                  className="validate"
                  name="title"
                  value={values.title}
                  onChange={changeHandler}
                />
                <label htmlFor="idea_title">Title *</label>
                <span className="helper-text">{errors.title}</span>
              </div>
              <div
                className={`description-field input-field col s12 ${
                  touched.description && errors.description
                    ? "invalid-field"
                    : ""
                }`}
              >
                <textarea
                  id="description"
                  className={`materialize-textarea ${
                    touched.description && errors.description ? "invalid" : ""
                  }`}
                  name="description"
                  onChange={changeHandler}
                  value={values.description}
                ></textarea>
                <label htmlFor="description">Description *</label>
                <span
                  className="character-counter"
                  style={{ float: "right", fontSize: "12px" }}
                >
                  {values.description.length}
                </span>
                <span className="helper-text">{errors.description}</span>
              </div>
              <div
                className={`tags-field input-field col s12 ${
                  touched.tags && errors.tags ? "invalid-field" : ""
                }`}
              >
                <select
                  multiple
                  onChange={changeHandler}
                  name="tags"
                  className={`${touched.tags && errors.tags ? "invalid" : ""}`}
                >
                  <option value="" disabled>
                    Select one or more tags.
                  </option>
                  <option value="feature">Feature</option>
                  <option value="tech">Tech</option>
                  <option value="bugfix">Bugfix</option>
                </select>
                <label>Tags</label>
                <span className="helper-text">{errors.tags}</span>
              </div>
            </form>
          </div>
        </div>
        <div className="modal-footer">
          <a
            className="cancel-btn modal-close waves-effect waves-red btn-flat"
            onClick={() => onCancel(false)}
          >
            Close
          </a>
          <button
            disabled={!isValid}
            className="submit-btn modal-close waves-effect waves-green btn-flat"
            onClick={handleSubmit}
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  );
};

export default RegisterDialog;
