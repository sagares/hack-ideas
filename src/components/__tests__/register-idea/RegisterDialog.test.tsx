/**
 * @jest-environment jsdom
 */
import React from "react";
import { act, fireEvent, render } from "@testing-library/react";
import { registerIdea } from "../../../utils/ServiceUtil";
import HackIdeasContext from "../../../context";
import { useForm } from "../../../hooks";
import RegisterDialog from "../../register-idea/RegisterDialog";

jest.mock("../../../utils/ServiceUtil");
//jest.mock("../../../hooks");

describe("Register Dialog Component", () => {
  let element, component, onCancel, changeHandler, dateNowSpy;

  beforeEach(() => {
    changeHandler = jest.fn().mockImplementation();
    dateNowSpy = jest
      .spyOn(Date, "now")
      .mockImplementation(() => 1487076708000);
    onCancel = jest.fn().mockImplementation(() => {});
    (registerIdea as jest.Mock).mockImplementation(() => {});

    element = document.createElement("div");
    component = render(
      <HackIdeasContext.Provider
        value={{
          currentUser: { empId: "123456" },
        }}
      >
        <RegisterDialog isOpen={true} onCancel={onCancel} />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );
  });

  afterEach(() => {
    document.body.removeChild(element);
  });

  test("renders", () => {
    expect(component.container).toMatchSnapshot();
  });

  test("enables submit button if form is valid", () => {
    const titleInput = element.querySelector('input[name="title"]');
    const description = element.querySelector('textarea[name="description"]');
    const tags = element.querySelector('select[name="tags"]');

    fireEvent.change(titleInput, { target: { value: "mock title" } });
    fireEvent.change(description, { target: { value: "mock description" } });
    fireEvent.change(tags, { target: { value: "feature" } });

    const submitButton = element.querySelector(".modal-footer .submit-btn");
    expect(submitButton.disabled).toBe(false);
  });

  test("submits idea on click of submit button", () => {
    const titleInput = element.querySelector('input[name="title"]');
    const description = element.querySelector('textarea[name="description"]');
    const tags = element.querySelector('select[name="tags"]');

    fireEvent.change(titleInput, { target: { value: "mock title" } });
    fireEvent.change(description, { target: { value: "mock description" } });
    fireEvent.change(tags, { target: { value: "feature" } });

    const submitButton = element.querySelector(".modal-footer .submit-btn");
    fireEvent.click(submitButton);

    expect(registerIdea).toHaveBeenCalledWith({
      description: "mock description",
      name: "mock title",
      submittedBy: "123456",
      submittedOn: 1487076708000,
      tags: ["feature"],
    });
    expect(onCancel).toHaveBeenCalledWith(false);
  });

  test("disables submit button if form is invalid", () => {
    const submitButton = element.querySelector(".modal-footer .submit-btn");
    expect(submitButton.disabled).toBe(true);
  });

  test("closes modal on click of cancel button", () => {
    const cancelButton = element.querySelector(".modal-footer .cancel-btn");
    fireEvent.click(cancelButton);
    expect(onCancel).toHaveBeenCalledWith(false);
  });

  test("shows validation messages for form fields", () => {
    const titleInput = element.querySelector('input[name="title"]');
    const description = element.querySelector('textarea[name="description"]');
    const tags = element.querySelector('select[name="tags"]');

    fireEvent.change(titleInput, { target: { value: "mock title" } });
    fireEvent.change(description, { target: { value: "mock description" } });
    fireEvent.change(tags, { target: { value: "feature" } });

    fireEvent.change(titleInput, { target: { value: "" } });
    fireEvent.change(description, { target: { value: "" } });
    fireEvent.change(tags, { target: { value: null } });

    const titleMessage = element.querySelector(".title-field .helper-text");
    const descriptionMessage = element.querySelector(
      ".description-field .helper-text"
    );

    expect(titleMessage.innerHTML).toBe("Title is required");
    expect(descriptionMessage.innerHTML).toBe("Description is required");
  });

  test("closes modal when isOpen is false", () => {
    component = render(
      <HackIdeasContext.Provider
        value={{
          currentUser: { empId: "123456" },
        }}
      >
        <RegisterDialog isOpen={false} onCancel={onCancel} />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );

    expect(
      component.container
        .querySelector(".register-idea--modal")
        .classList.contains("open")
    ).toBe(false);
  });
});
