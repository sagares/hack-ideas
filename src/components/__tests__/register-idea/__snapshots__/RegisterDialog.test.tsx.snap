// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Dialog Component renders 1`] = `
<div>
  <div
    class="register-idea--modal open"
  >
    <div
      class="modal modal-fixed-footer"
      id="modal1"
    >
      <div
        class="modal-content"
      >
        <h4>
          Register New Idea
        </h4>
        <p>
          Please fill in the form below to register your hack idea.
        </p>
        <div
          class="row"
        >
          <form
            class="col s12"
          >
            <div
              class="title-field input-field col s12 "
            >
              <input
                class="validate"
                id="idea_title"
                name="title"
                type="text"
                value=""
              />
              <label
                for="idea_title"
              >
                Title *
              </label>
              <span
                class="helper-text"
              />
            </div>
            <div
              class="description-field input-field col s12 "
            >
              <textarea
                class="materialize-textarea "
                id="description"
                name="description"
              />
              <label
                for="description"
              >
                Description *
              </label>
              <span
                class="character-counter"
                style="float: right; font-size: 12px;"
              >
                0
              </span>
              <span
                class="helper-text"
              />
            </div>
            <div
              class="tags-field input-field col s12 "
            >
              <div
                class="select-wrapper"
              >
                <input
                  class="select-dropdown dropdown-trigger"
                  data-target="select-options-bacaaf9f-26b1-244a-c760-92f9c2b9d088"
                  readonly="true"
                  type="text"
                />
                <ul
                  class="dropdown-content select-dropdown multiple-select-dropdown"
                  id="select-options-bacaaf9f-26b1-244a-c760-92f9c2b9d088"
                  tabindex="0"
                >
                  <li
                    class="disabled"
                    id="select-options-bacaaf9f-26b1-244a-c760-92f9c2b9d0880"
                    tabindex="0"
                  >
                    <span>
                      <label>
                        <input
                          "=""
                          disabled=""
                          type="checkbox"
                        />
                        <span>
                          Select one or more tags.
                        </span>
                      </label>
                    </span>
                  </li>
                  <li
                    id="select-options-bacaaf9f-26b1-244a-c760-92f9c2b9d0881"
                    tabindex="0"
                  >
                    <span>
                      <label>
                        <input
                          "=""
                          type="checkbox"
                        />
                        <span>
                          Feature
                        </span>
                      </label>
                    </span>
                  </li>
                  <li
                    id="select-options-bacaaf9f-26b1-244a-c760-92f9c2b9d0882"
                    tabindex="0"
                  >
                    <span>
                      <label>
                        <input
                          "=""
                          type="checkbox"
                        />
                        <span>
                          Tech
                        </span>
                      </label>
                    </span>
                  </li>
                  <li
                    id="select-options-bacaaf9f-26b1-244a-c760-92f9c2b9d0883"
                    tabindex="0"
                  >
                    <span>
                      <label>
                        <input
                          "=""
                          type="checkbox"
                        />
                        <span>
                          Bugfix
                        </span>
                      </label>
                    </span>
                  </li>
                </ul>
                <svg
                  class="caret"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
                <select
                  class=""
                  multiple=""
                  name="tags"
                  tabindex="-1"
                >
                  <option
                    disabled=""
                    value=""
                  >
                    Select one or more tags.
                  </option>
                  <option
                    value="feature"
                  >
                    Feature
                  </option>
                  <option
                    value="tech"
                  >
                    Tech
                  </option>
                  <option
                    value="bugfix"
                  >
                    Bugfix
                  </option>
                </select>
              </div>
              <label>
                Tags
              </label>
              <span
                class="helper-text"
              />
            </div>
          </form>
        </div>
      </div>
      <div
        class="modal-footer"
      >
        <a
          class="cancel-btn modal-close waves-effect waves-red btn-flat"
        >
          Close
        </a>
        <button
          class="submit-btn modal-close waves-effect waves-green btn-flat"
          disabled=""
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
